File: gamedex-common/src/main/java/com/github/ykrasik/gamedex/common/util/UrlUtils.java
package com.github.ykrasik.gamedex.common.util;

import com.github.ykrasik.opt.Opt;
import com.google.common.io.Resources;
import lombok.SneakyThrows;

import java.awt.*;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.net.URI;
import java.net.URL;

/**
 * @author Yevgeny Krasik
 */
public final class UrlUtils {
    private UrlUtils() { }

    public static Opt<byte[]> fetchOptionalUrl(Opt<String> url) throws IOException {
        return url.flatMapX(UrlUtils::fetchData);
    }

    public static Opt<byte[]> fetchData(String urlStr) throws IOException {
        final URL url = new URL(urlStr);
        try {
            return Opt.of(Resources.toByteArray(url));
        } catch (FileNotFoundException e) {
            return Opt.absent();
        }
    }

    @SneakyThrows
    public static void browseToUrl(String url) {
        Desktop.getDesktop().browse(new URI(url));
    }
}

File: gamedex-common/src/main/kotlin/com/github/ykrasik/gamedex/common/util/UrlUtils.kt
package com.github.ykrasik.gamedex.common.util

import com.google.common.io.Resources
import java.awt.Desktop
import java.io.FileNotFoundException
import java.net.URI
import java.net.URL

/**
 * User: ykrasik
 * Date: 02/10/2016
 * Time: 22:45
 */

fun String.fetchUrlData(): ByteArray? {
    val url = URL(this)
    return try {
        Resources.toByteArray(url)
    } catch (e: FileNotFoundException) {
        null
    }
}

fun browseToUrl(url: String) {
    Desktop.getDesktop().browse(URI(url))
}